### PR TITLE
accounts: make KDFParams a struct

### PR DESCRIPTION
**Pull request by htkao**

### CODE REVIEW

File: accounts/keystore/passphrase.go [hunk 1]
```diff
@@ -319,38 +321,28 @@ func decryptKeyV1(keyProtected *encryptedKeyJSONV1, auth string) (keyBytes []byt
...
+			cryptoJSON.KDFParams.P,
+			cryptoJSON.KDFParams.DKLen,
+		)
 
 	} else if cryptoJSON.KDF == "pbkdf2" {
-		c := ensureInt(cryptoJSON.KDFParams["c"])
-		prf := cryptoJSON.KDFParams["prf"].(string)
-		if prf != "hmac-sha256" {
-			return nil, fmt.Errorf("Unsupported PBKDF2 PRF: %s", prf)
+		if cryptoJSON.KDFParams.Prf != "hmac-sha256" {
```
Comment (holiman) [inline accounts/keystore/passphrase.go]: AFAICT, this is a functional change. Previously, we would not validate that `c` is present / is a number, unless `pbkdf2` was used. With this change, the json parser will reject `scrypt`:ed keystores which does not have `c`. I might be wrong, but we need to ensure that we have good test coverage here, perhaps testcases that include keystores generated from python (web3.py), javascript, parity, as well as both version 1 and version 3, to ensure that we don't break parsing of old or somewhat malformed keystores which we were able to decode previously.

Comment (htkao) [reply accounts/keystore/passphrase.go]: Great point. I agree that we shouldn't validate whether the fields are present so I added "omitempty" to the struct. The json parser also won't reject keystore which contains parameter that's not defined in the struct. 

In this patch, field in "kdfparamsJSON" will get rejected if data type is wrong. e.g. `cryptoJSON.KDF == scrypt` and `c` is a string. In my mind, we should reject this case although `c` is not being used. What do you think?

Comment (holiman) [reply accounts/keystore/passphrase.go]: Well, it's better (less breaking), but I'm not sure about this. If we make it a struct, we paint ourselves into a corner where we have to add `omitempty` or special formatting directives in the json parser, which quickly becomes tricky. And I'm not sure we have sufficient test-coverage to ensure that we don't break odd variants.

**Comment by holiman** [comment]
Thanks for the work you have put in, but I would prefer to reject this PR, since it might break things and I don't really see a big win in 'structifying' these things at this point. 
I'll leave it open for a while longer to see if anyone else opines.

**Pull request closed without merging**